### ISSUE TITLE
Fragmented faces in texture mapping cause blurriness and texture error

### ISSUE THREAD

**Issue by otepmong** [cdcseacave/openMVS#331]
Hello,

In the OpenMVS TextureMesh program, we noticed many texture regions of the generated model containing either a small number of faces (or a single face), or an 'island' where a group of faces are mapped to one image, but all surrounding faces are mapped to a different image. Please see the screenshot below, on the left, using MeshLab to show texture seams.

![oldnewbluelaxt](https://user-images.githubusercontent.com/7981417/40687374-46a740e4-634f-11e8-86c9-46ff691b0e19.png)

We have been working on an option to remap these islands to the surrounding texture region by modifying the label selections after labels are extracted from the inference step in SceneTexture.cpp:FaceViewSelection(), but before texture patches are actually created in GenerateTexture(). We've had reasonable success (right side of screenshot).

Additional improvements we are working on require that for a given face, finding all of its valid labels. We were thinking of iterating through the face's data in FaceDataViewArr facesDatas, but are not sure how a view ID and a face's label are connected (if at all). We would greatly appreciate If you have any hints/pointers to the class relationships!

Also, we would be happy to submit the code back to you if you think it would be useful to others.

Thank you in advance..

Oh Tepmongkol-Wheaton

### Specifications like the version of the project, operating system, and hardware
System: Ubuntu 16.04 LTS
OpenMVG: v1.1-329-gdb92617  ( db92617fea420363abcbad1cb3773706f75012ad )
OpenMVS: v0.7-22-g939033c ( 939033c55b50478339084431aac2c2318041afad )

### Steps to reproduce the problem
Command lines to reproduce:
1) openMVG_main_SfMInit_ImageListing -i ../Images -o ./mvgMatch -d sensor_width_camera_database.txt -k '2.5741200457829959e+03;0;9.6820590653514193e+02;0.0; 2.5741200457829959e+03;1.2809532405422367e+03;0.0;0.0;1.0'
2) openMVG_main_ComputeFeatures -i ./mvgMatch/sfm_data.json -o ./mvgMatch -f 1 -m AKAZE_FLOAT -p HIGH 
3) openMVG_main_ComputeMatches -i ./mvgMatch/sfm_data.json -o ./mvgMatch -f 1 
4) openMVG_main_IncrementalSfM -i ./mvgMatch/sfm_data.json -m ./mvgMatch -o ./mvgReconstruction 
5) openMVG_main_ConvertSfM_DataFormat -i ./mvgReconstruction/sfm_data.bin -I -V -E -o ./mvgReconstruction/sfm_data_cameras.json 
6) DensifyPointCloud --input-file ./mvs/scene.mvs --estimate-normals 1 --resolution-level 1 
7) ReconstructMesh ./mvs/scene_dense.mvs 
8) RefineMesh ./mvs/scene_dense_mesh.mvs --resolution-level 1 --max-views 8 --scales 3 --decimate 0.10 
9) TextureMesh ./mvs/scene_dense_mesh_refine.mvs --cost-smoothness-ratio 1 --export-type obj --resolution-level 0

**Comment by cdcseacave** [comment]
Hi @otepmong 

Thank you for the great topic! A wile back I also tried to solve this issue, but gave up shortly after without success. My approach was to try solving this right at the source, fine-tuning the edge weights and the inference method. In the short time I allocated for it I didn't find a way to reduce the number of islands, and if I remember right I was intrigued of the constancy of the step, as almost any changes I would make did not affect much the final outcome.

Anyway, going back to your approach, I like it too, the only thing is to be careful to reassign a view that sees that island of faces, preferably more or less as well as the current one. All you need it to look at `FaceDataViewArr facesDatas;`. That structure contains for each face the list of images seeing it (`idxView`), and how well the view sees it (`quality`). The label use during inference step is `idxView` plus an extra label for not-assigned.

If you have the time I suggest to have a second look at the quality and inference step and try solving this issue there, which would be preferable over the post-processing reassignment.

**Comment by raisindoc** [comment]
Hi @cdcseacave ,

I'm Alex, I'm working with Oh on the code; based on what you suggested, we added a validation step to make sure that if we changed the label of a face, the new label was valid in its corresponding FaceDataArr in facesDatas. The results we found are below:
![validatestep](https://user-images.githubusercontent.com/25215489/40742798-7827587e-6404-11e8-90a3-8a01cde44497.png)
The left is what we had before without the validation step, the middle is without any changes to the original open source code, and the right is after we added the validation step for each face. For our purposes, it seems that for an 'island', the surrounding label should also be valid for the island, but this does not seem to be true for many of the island faces; we think that when facesDatas is initialized, some potentially valid views for a face may be ignored. Any idea what might be causing this, or where in the code we should look to try to change it?

Thank you in advance,
Alex Wang

(Specifications and steps are the same as Oh's)

**Comment by cdcseacave** [comment]
This is very interesting, I think you found something important here!

Here is how the views are assigned to each face:
- an octree is filled with all faces
- for each view
  - based on the octree, select all faces that are inside the view frustrum
  - render the visible face in this view and keep for each pixel the face ID that project there
  - all faces that have an ID in the projected face-map are set as visible from this view

So a face is visible in a view if that face is in the view frustrum, is oriented towards that view, and is not behind an other face. And so first of all I suggest to check if a face is really visible in the desired view by verifying all the previous constraints. If a face is really visible in a view but it is not marked as visible looking at this, there are few steps that can be the problem, but the most likely is that that face is so small that it does not occupy not even a pixel when projected in the image. There are few ways to solve this, but pls first identify the problem.

**Comment by raisindoc** [comment]
Thank you so much for your response! We dug a little further into the issue and found the following, examining one of the outlier faces which had this problem:
![missingidxview](https://user-images.githubusercontent.com/25215489/40805245-801e815a-64d2-11e8-8c15-816d4c8bec5a.png)
We looked at the highlighted (small orange) face on the left below the 'q' and 'u' and found that the image used for the texture of this face was the one on the middle, while the image used for the large patch above it was the one on the right; this face seems to be well within view of the right image, but this view was not listed for that particular face. The size of the face in relation to pixels doesn't seem to be the issue: this face is around the same size as the faces around it, all of which are significantly larger than a pixel. 
We're unsure how to proceed from this point. Might the issue be from the lighting pattern we have or some other issue with our images, or can we try fixing it somewhere in the code?

Thank you in advance,
Alex

**Comment by raisindoc** [comment]
Thank you! We did some debugging over the weekend and found that views would generally be identified correctly but would be removed from facesDatas during outlier detection. We experimented a bit with changing the outlier threshold parameter and found that results would improve as we made it smaller, up to a certain point where the resulting model would have very obvious errors. To our understanding, outlier detection helps in cases where part of a view is obstructed (like with passerby in other applications) but for our purposes, there are no obstructions in our images, so we don't understand why removing outlier detection negatively affects the result so much; would you be willing to share any insights as to why this might be the case, or how we might tackle this problem?

Thank you in advance,
Alex

**Comment by cdcseacave** [comment]
Inaccuracy in camera calibration are in general of many reasons, but you seem to try to reconstruct a human, which if a person it might slightly move from picture to picture. It is also possible that the camera is precise enough, but the mesh is not perfect, so there are small miss-alignments when projected into the images. Hard to say without any data. Depends on that "small" means. If very small, maybe that is the maximum accuracy given the data, and we might need to find a way to deal with that. In that case maybe your initial solution of reassigning the face to the surrounding view is a good solution.

**Comment by otepmong** [comment]
Thanks for all your feedback!

Hopefully this image can help explain what we are seeing and our current theories. In our setup, most reconstructed faces are viewable from 20-30 images/camera positions (though some of them being at extreme angles). The left image is the final texture mapped model, where a large, relatively flat mesh region is texture mapped from Image 1, while a single 'island' face is texture mapped from Image 2. For the four images on the right, the white triangle shows the position of the rasterized projected 'island' face on a few source images (using the rasterized faceMap data from SceneTexture.cpp:MeshTexture::ListCameraFaces()).

In this case, the face is close to a blue/pink boundary, and while the face falls more on blue than pink in Image 1, the face falls more on pink than blue on many other source images, which we believe is due to small errors in the calculated camera positions. Image 1 subsequently gets marked as an outlier and dropped because the mean color falls outside the threshold (which we lowered to 0.01 in this case). Image 4 is an example of an extreme angle projection, which we think may also throw off the mean color outlier check.

This is why we think that this issue may be caused by the camera position calculation not being accurate enough. We are currently looking at ways to improve this.

![rogue triangle and faces](https://user-images.githubusercontent.com/7981417/41168198-1168a3f2-6afa-11e8-86ef-346a19bf78c2.jpg)

**Comment by otepmong** [comment]
Just to re-iterate our setup.

1. We have an array of 107 cameras, all in fixed location - see image of camera poses as outputed from openMVG
2. the objects of interest at the center are inanimate, not human.  So neither camera or object is moving during photo.
3. Our cameras are of the same type with 5MP.  We think the quality is sufficient.  We ran the same set of images through Agisoft Photoscan Pro and its texture result did not have this issue.

![camera pose image - pierre](https://user-images.githubusercontent.com/7981417/41168654-718a73a4-6afb-11e8-938e-9fed715f343b.png)

**Comment by otepmong** [comment]
We uploaded the images.  Please follow the instructions 

https://drive.google.com/open?id=1HaJpD0XJdqVxHbCdq6Xy1YQjZMlTOERl

**Comment by otepmong** [comment]
For your reference, Agisoft PhotoScan 14.1 output.

![screenshot from 2018-06-08 16-39-23](https://user-images.githubusercontent.com/7981417/41185411-0038b9d0-6b3d-11e8-9d3f-0258ff2f653a.png)

**Comment by cdcseacave** [comment]
Professional work (like till now)! The slide pictures exactly the theory you had: there are imperfections in the camera pose and/or the mesh. Of course, the first think we can try is to improve those. In your setup there is one easy way to remove the camera poses from the equation by pre-calibrating your cameras and keep them fix for all future reconstructions. There are several ways to do that, the most common one is using a pattern for calibration and if you need I can share some code you can use for that based on some OpenCV routines.

However, there will always be imperfections, so we need to improve also the robustness of the texturing module. Here is an idea I had while reading your findings. Instead of using the ourlier removal for each face individually before the labeling inference, I think it is much more powerful to postpone that till the label inference, and incorporate the color (outlier) information as an additional weight for the quality weight used for inference. Let's call this new weight the color weight, and let's compute it as the std-dev from the mean color. We should combine this color weight with the quality weight such that the influence of the quality is still more important, and the color weight does not eclipse the smoothness weight. This way the inference should prefer to assign the view to a face that was assign to the surrounding faces, but still be robust in face of outliers. Can you pls try to implement this?

**Comment by cdcseacave** [comment]
I pushed the code I have in https://github.com/cdcseacave/openMVS/tree/multi_camera, it follows the documentation from here: https://docs.opencv.org/3.3.0/d2/d1c/tutorial_multi_camera_main.html Pls tell me if you find it useful or you need improvements.

**Comment by otepmong** [comment]
@cdcseacave  We were able to build the Multi Camera Calibration, by building the whole OpenMVS.  Can I build only Multi Camera Calibration ?  Please send instructions.  Thank you.

**Comment by cdcseacave** [comment]
I am not sure I understand, Multi Camera Calibration is part of OpenMVS and it will be build together with OpenMVS; you can run it as you run any app

**Comment by otepmong** [comment]
We have a small laptop we operate the camera arrays and perform intrinsic calibration (opencv).  Ideally, we want to run Multi Camera Calibration on the same machine as well if it is the same order of computation and only opencv library is used.  I don't believe we can build OpenMVS on this machine - not enough juice.  Our OpenMVG/OpenMVS software is running on a powerful server.  Hopefully, this makes sense.

**Comment by cdcseacave** [comment]
But you can build OpenMVS wherever you want, and just use the binaries on your laptop (simply copy the EXE files). However the calibration is computationally demanding as well, as it does a lot of feature extraction and matching.

**Comment by cdcseacave** [comment]
The normal camera calibration procedure should be employed for orientation and distance: preferably the pattern should cover as much as possible the image, from all sides, and at different scales. However the patterns does not need to be visible entirely in one image to be used, partial is enough, but the more the better. For more details follow the tutorial of from the original paper: https://sites.google.com/site/prclibo/toolbox/doc

Pls note that I never had a chance to test the code, never had access to a multi camera setup.

**Comment by otepmong** [comment]
It has been a while.  Happy New Year!

We were not able to come up with a reliable solution for this issue.  We improve our setup to reduce the occurrences of these isolate island. 

Please close this issue.

**Comment by thunders82** [comment]
Hi @otepmong 
I was wondering if you have shared your code to remap the single isolated faces to bigger islands ?
I'm facing similar issue and maybe your code could help me.
Thank you

**Comment by tftangming** [comment]
@otepmong  Hello! I meet the issue also. And I want to know if you have specific methods to solve the problem? What I can know is to search the islands which own different label A from its surrounding components faces with same label B firstly, and then modify the label A to B? But  I think it is time-consuming too much during finding these islands.

**Comment by otepmong** [comment]
We did not find a reliable solution for this issue.  We were able to reduce these isolated islands by improving our camera calibration per @cdcseacave suggestion.   Since we are in a very controlled setting, we dont need to worry about transient occlusions and turn off this function.  In general openmvs handling transient occlusions is an important feature.  I will send a summary of what we have tried.
